rfcs/0006-gtfs-canonical-routes.md

```diff
@@ -0,0 +1,108 @@
+- **shapes.txt** can probably use the existing shapes.
+- **trips.txt** includes one new trip for each new route pattern. It should reference aforementioned
+  route pattern and shapes. It should reference the new `service_id` created above. Each of these
+  trips should be the `representative_trip_id` for their respective route patterns.
+- **stop_times.txt** includes all stops that trains in the route/direction passes through, even if
+  they are temporarily closed. Increment the arrival/departure times by one minute for each
+  subsequent stop. For a sample, see the attached GTFS file, the output of [✓ 🧪 🚝 Canonical
+  trips/patterns experiment](https://app.asana.com/0/881264583703207/1200210504369250), which
+  trialled this for Green Line D and Lowell Line (only in the outbound direction).
+
+This will be done as a new "step" of the Makefile called `canonical`, which runs between `truncate`
+and `current`. There will be a new a directory, `input/canonical/` to store any static data the step
+uses.
+
+The canonical step will be a Python module that reads in the `truncate` feed, and writes out its own
+feed after transforming it. The transformations are as above, and also will update any current trips
+to use canonical route patterns if relevant, eliminating redundant route patterns.
+
+For Commuter Rail route patterns, the canonical one will specify every stop as non-flag, non-leave
+early stop.
+
+# Drawbacks
+
+This is more complicated than not doing it. Some apps may prefer to treat the "typical" route
+pattern as the one that has current service. For example, in the case of a long-term stop closure,
+perhaps some apps don't want to show the closed stop at all.
+
+# Rationale and alternatives
+
+We could leave the current route patterns and trips alone, and add canonical route patterns with a
+new typicality value of `5`. Their only trips would be their single representative trip. With this
+approach, since there's no interaction with the remaining data at all, we would make the `canonical`
+feed one that goes into the `merged` step. (Though we'd have to figure out how to determine the
+valid feed start and end date.)
+
+# Prior art
+
+Until now we've been doing ad-hoc adjustments to determine the route pattern of routes
+[here](https://github.com/mbta/api/blob/master/apps/state/config/config.exs#L145).
+
+# Unresolved questions
+
+The main question is whether current service should be put on canonical route patterns, if
+applicable, or not.
+
+# Future possibilities
+
+We could expose this more clearly in the API. In the approach outlined here, the "canonical" route
+pattern is simply that with a typicality of `1`. In the alternative approach, the typicality would
+be `5`. That's not currently filterable in the API. In addition, the `service` can't be filtered by
+its typicality, most relevantly when included with `trips`.
+
+Shoudl we do something similar for ferry and/or bus?
```

Comment:
Future v3 idea related to bus (if we don't have canonical trips stored for every bus route): Whenever we add a ["HASTUS patch" for bus](https://github.com/mbta/gtfs_creator/blob/master/gtfs_creator2/hastus/update.py#L1243), we automatically create a canonical route pattern with the pre-patched stop list.